### PR TITLE
[DAG] isConstantIntBuildVectorOrConstantInt - peek through bitcasts

### PR DESCRIPTION
**Pull request by RKSimon**
Alter both isConstantIntBuildVectorOrConstantInt + isConstantFPBuildVectorOrConstantFP to return a bool instead of the underlying SDNode, and adjust usage to account for this.

Update isConstantIntBuildVectorOrConstantInt to peek though bitcasts when attempting to find a constant, in particular this improves canonicalization of constants to the RHS on commutable instructions.

X86 is the beneficiary here as it often bitcasts rematerializable 0/-1 vector constants as vXi32 and bitcasts to the requested type

Minor cleanup that helps with #107423

Alternative to #112682

### CODE REVIEW

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 1]
```diff
@@ -56552,8 +56552,8 @@ static SDValue combineSub(SDNode *N, SelectionDAG &DAG,
 
   // TODO: Add NoOpaque handling to isConstantIntBuildVectorOrConstantInt.
   auto IsNonOpaqueConstant = [&](SDValue Op) {
-    if (SDNode *C = DAG.isConstantIntBuildVectorOrConstantInt(Op)) {
-      if (auto *Cst = dyn_cast<ConstantSDNode>(C))
+    if (DAG.isConstantIntBuildVectorOrConstantInt(Op)) {
+      if (auto *Cst = dyn_cast<ConstantSDNode>(Op))
```
[the remark this answers is not in the document]
Comment (RKSimon) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:56556]: good catch!

**Review by phoebewang — APPROVED** [review]
LGTM.

**Comment by mstorsjo** [comment]
This change broke compilation for Windows/ARM - some files now make the compiler to not terminate (running for nearly forever, and possibly consuming all memory).

This can be reproduced with https://martin.st/temp/llvm-preproc.cpp, compiled as:
```
$ clang -target armv7-w64-mingw32 -c -O2 llvm-preproc.cpp
```
Previously, this finished compiling in a matter of seconds, now it doesn't finish.

Pushing a revert shortly.

**Comment by RKSimon** [comment]
@mstorsjo Recommitted the commit - should be all good now!

**Comment by mstorsjo** [comment]
> @mstorsjo Recommitted the PR - should be all good now!

Yes, looking good now - thanks!

**Pull request merged**